### PR TITLE
#356 #695 #1924 Custom `HttpMessageInvoker` pooling

### LINKED ISSUE #356: Implementation with HttpClientFactory in ASP.NET Core 2.1
Hey Tom! - Are you planning to move to **HttpClientFactory** in **ASP.NET Core 2.1** when moving Ocelot to .NET Core 2.1 in the upcoming future? :)

See a summary from Glenn in my team:
https://blogs.msdn.microsoft.com/webdev/2018/02/28/asp-net-core-2-1-preview1-introducing-httpclient-factory/

### Update on Nov 29, 2023
- [IHttpClientFactory with .NET](https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory)
- [Make HTTP requests using IHttpClientFactory in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-requests)
- [Use IHttpClientFactory to implement resilient HTTP requests](https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests)
- Bing.com: [Search for "IHttpClientFactory"](https://www.bing.com/search?q=IHttpClientFactory)

### PR DESCRIPTION
**Pull request by ggnaegi**
## Closes #356 #695 #1924
- #356 
- #695 
- #1924

After some checks, it seems impossible to use `HttpClientFactory ` in the solution. 
There are several reasons for this: 
- First, the `HttpMessageHandler ` can potentially be unique for each route. In the case of a "Named" `HttpClient`, it's possible to use `ConfigurePrimaryHttpMessageHandler`, but this method will only be invoked the first time `HttpClientFactory.CreateClient("Named")` is called. In our case, this would mean that we are configuring as many "Named" clients as there are routes. This is not quite what we are looking for. 
- Second point, the use of resources. It seems that `HttpMessageInvoker ` is potentially less resource-intensive than `HttpClient`. Of course, `HttpMessageInvoker ` is not returned by `HttpClientFactory ` and worse, if we force the cast, then we could face major performance problems.
  See: https://github.com/microsoft/reverse-proxy/issues/458

I conclude that we must implement our own pool mechanism...
Thus, I propose a first draft of a solution using solutions borrowed from StackExchange and Microsoft. It will obviously have to be tested in detail, but it serves as a basis for discussion.

## Proposed Changes
  - Using `HttpMessageInvoker` instead of `HttpClient`
  - Using `SocketsHttpHandler`, introducing a new parameter `PooledConnectionLifetime` in `HttpHandlerOptions`
  - Proposing an implementation of `IMessageInvokerPool`, the pool of message invokers, inspired by StackExchange [implementation](https://github.com/StackExchange/StackExchange.Utils/blob/main/src/StackExchange.Utils.Http/DefaultHttpClientPool.cs)
  - Removing redundant classes used in Ocelot (`HttpClientBuilder`, `HttpClientWrapper` and others)

I will provide Acceptance and Unit tests later. You should @raman-m, @RaynaldM have a look at this draft. Thanks

| payloadName         | Mean           | Max            | Op/s       | Allocated      |
|---------------------|----------------|----------------|------------|----------------|
| 1KBPayload.json     | 591.3 us       | 648.3 us       | 1,691.1792 | 39.78 KB       |
| 16KBPayload.json    | 642.4 us       | 664.7 us       | 1,556.7119 | 82.36 KB       |
| 32KBPayload.json    | 650.1 us       | 675.2 us       | 1,538.2294 | 192.68 KB      |
| 64KBPayload.json    | 827.9 us       | 841.7 us       | 1,207.8715 | 350.25 KB      |
| 128KBPayload.json   | 997.5 us       | 1,018.1 us     | 1,002.5522 | 668.68 KB      |
| 256KBPayload.json   | 1,424.4 us     | 1,461.5 us     | 702.0456   | 1303.03 KB     |
| 512KBPayload.json   | 2,323.8 us     | 2,466.2 us     | 430.3273   | 2568.97 KB     |
| 2048KBPayload.json  | 8,299.8 us     | 8,570.5 us     | 120.4855   | 10171.14 KB    |
| 8192KBPayload.json  | 24,469.6 us    | 25,469.2 us    | 40.867     | 40619.62 KB    |
| 10MBPayload.dat     | 29,709.8 us    | 32,654.6 us    | 33.6589    | 42708.54 KB    |
| 15360KBPayload.json | 54,608.4 us    | 70,385.4 us    | 18.3122    | 47989.27 KB    |
| 30720KBPayload.json | 101,482.4 us   | 114,551.1 us   | 9.8539     | 95953.63 KB    |
| 100MBPayload.dat    | 381,869.0 us   | 419,428.2 us   | 2.6187     | 362479 KB      |
| 1024MBPayload.dat   | 3,300,482.0 us | 3,444,481.6 us | 0.303      | 5193381.84 KB  |

Table 1: Performance with current http client 

| payloadName         | Mean           | Max            | Op/s       | Allocated      |
|---------------------|----------------|----------------|------------|----------------|
| 1KBPayload.json     | 579.6 us       | 588.2 us       | 1,725.2972 | 39.26 KB       |
| 16KBPayload.json    | 619.0 us       | 666.1 us       | 1,615.4381 | 66.49 KB       |
| 32KBPayload.json    | 661.8 us       | 670.4 us       | 1,511.0611 | 160.79 KB      |
| 64KBPayload.json    | 856.2 us       | 888.3 us       | 1,167.9673 | 286.51 KB      |
| 128KBPayload.json   | 886.4 us       | 901.6 us       | 1,128.1035 | 540.97 KB      |
| 256KBPayload.json   | 1,478.0 us     | 1,503.4 us     | 676.5674   | 1047.99 KB     |
| 512KBPayload.json   | 2,539.9 us     | 2,570.8 us     | 393.713    | 2061.79 KB     |
| 2048KBPayload.json  | 8,232.4 us     | 9,159.4 us     | 121.4709   | 8128.46 KB     |
| 8192KBPayload.json  | 25,601.3 us    | 26,334.2 us    | 39.0605    | 32306.69 KB    |
| 10MBPayload.dat     | 26,567.3 us    | 27,575.2 us    | 37.6403    | 32300.54 KB    |
| 15360KBPayload.json | 29,534.4 us    | 31,930.2 us    | 33.8588    | 32361.59 KB    |
| 30720KBPayload.json | 100,626.1 us   | 103,409.9 us   | 9.9378     | 64746.37 KB    |
| 100MBPayload.dat    | 330,750.7 us   | 351,597.2 us   | 3.0234     | 258272.69 KB   |
| 1024MBPayload.dat   | 3,441,790.4 us | 3,559,791.2 us | 0.2905     | 4124999.17 KB  | 

Table 2: Performance with the custom message invoker pool

### CODE REVIEW

File: src/Ocelot/Configuration/Creator/HttpHandlerOptionsCreator.cs
Comment (ggnaegi) [reply src/Ocelot/Configuration/Creator/HttpHandlerOptionsCreator.cs:10]: Yes, or we can foresee the new variable in global config...

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: Yes, or we can foresee the new param in global config...

Comment (RaynaldM) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: Sure @ggnaegi but we must stay focus on the httpclient issue, add some param in global could be done in a next issue

Comment (ggnaegi) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: ok

**Comment by raman-m** [comment]
Guys, Let's return to this awesome PR after [Nov'23](https://github.com/ThreeMammals/Ocelot/milestone/2) release.. 🆗 ❔

**Comment by ggnaegi** [comment]
> @raman-m @RaynaldM I'm writing the test cases at the minute. No worries guys! I realized we can still simplify the logic, since the Downstreamroute is an immutable, no need to use the timeout as key parameter, the object reference is good enough. We might have problems in the future if allowing dynamic configuration though (1: empty the pool, then 2: load the new configuration).

By the way, why not use the Record type instead?

**Comment by RaynaldM** [comment]
> > @raman-m @RaynaldM I'm writing the test cases at the minute. No worries guys! I realized we can still simplify the logic, since the Downstreamroute is an immutable, no need to use the timeout as key parameter, the object reference is good enough. We might have problems in the future if allowing dynamic configuration though (1: empty the pool, then 2: load the new configuration).
> 
> By the way, why not use the Record type instead?

It is a good idea

**Comment by ggnaegi** [comment]
Hello @raman-m and @RaynaldM I'm providing some benchmarks, it seems the memory leak has been addressed

**Comment by ggnaegi** [comment]
@raman-m Interesting: https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#large-object-heap-threshold

**Comment by raman-m** [comment]
Please, resolve the conflict!

**Comment by raman-m** [comment]
@ggnaegi As you may know, I've attached #695 ... Is it right?
Do you think that #695 will also be fixed by this PR?

----
The 695 quote:
> I couldn't find any configuration setting to enable request or response streaming.

What about this?

**Comment by raman-m** [comment]
Ready for code review

**Comment by ggnaegi** [comment]
@raman-m I think, it's the request mapper implementation that solved the issue #695 (we were copying the full request body to a memory stream). Now, with the http client pool we avoid using the http client full buffering, but as others wrote, the performance improvement is minimal.

**Review by RaynaldM — APPROVED** [review]

File: src/Ocelot/Errors/QoS/RequestTimedOutError.cs
Comment (raman-m) [inline src/Ocelot/Errors/QoS/RequestTimedOutError.cs]: Do we override 408 RequestTimeout here?

File: src/Ocelot/Requester/MessageInvokerHttpRequester.cs
Comment (raman-m) [inline src/Ocelot/Requester/MessageInvokerHttpRequester.cs:34]: Ugly mappers again... 😢

**Review by raman-m — APPROVED** [review]
Reviewed! Approved! ✅

**Comment by raman-m** [comment]
Ready for delivery! ✔️

File: src/Ocelot/Errors/QoS/RequestTimedOutError.cs
Comment (ggnaegi) [reply src/Ocelot/Errors/QoS/RequestTimedOutError.cs]: good point... I just moved the Error class from the polly provider to the base project. I don't know, what should we return, 408 or 503?

File: src/Ocelot/Requester/MessageInvokerHttpRequester.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/MessageInvokerHttpRequester.cs:34]: At the end the mappers are not that bad, but it's the way we are injecting them that is annoying.

**Comment by ggnaegi** [comment]
@raman-m @RaynaldM I would like to add a few test cases, ok?

File: src/Ocelot/Errors/QoS/RequestTimedOutError.cs
Comment (ggnaegi) [reply src/Ocelot/Errors/QoS/RequestTimedOutError.cs]: @raman-m Should we keep it like that for the time being and create a new issue later?

Comment (raman-m) [reply src/Ocelot/Errors/QoS/RequestTimedOutError.cs]: Returning **5xx** status is very bad in my opinion.
In addition we should think in terms of returning correct native status code from downstream, at least try to forward this native status.
But this is QoS logic where we try to override behavior of internal Polly features.
If it is possible it is better to return original status.
My concern was/is Why custom Error object? Cannot we just return exception or real result object.
Or this Error object plays the role of the result?

**Comment by raman-m** [comment]
> @ggnaegi commented [yesterday](https://github.com/ThreeMammals/Ocelot/pull/1824#issuecomment-1883036734)

Need to review once again?
I will review before the merge definitely! 😉 
Let us know when ready to review plz!

File: src/Ocelot/Requester/MessageInvokerHttpRequester.cs
Comment (raman-m) [reply src/Ocelot/Requester/MessageInvokerHttpRequester.cs:34]: Not an issue for this PR for sure. As you know we have to do something with this design of Errors Mapper! Reason is "**wrong wrapping & overriding original status codes**". That's very bad!
Could you add some TODO sentence to these lines with a comment to pay attention lates please.
We have to return to this problem once again after Annual release...

**Comment by ggnaegi** [comment]
@raman-m could we push this to the november release? I think it's high priority.

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyQoSProvider.cs:83]: Ah-ha-h!.. Do we have `Optimistic` strategy? 😄

File: src/Ocelot/Configuration/Creator/HttpHandlerOptionsCreator.cs
Comment (raman-m) [inline src/Ocelot/Configuration/Creator/HttpHandlerOptionsCreator.cs:25]:
```suggestion
            var pooledConnectionLifetime = TimeSpan.FromSeconds(options.PooledConnectionLifetimeSeconds ?? DefaultPooledConnectionLifetimeSeconds);
```

File: src/Ocelot/Configuration/HttpHandlerOptions.cs
Comment (raman-m) [inline src/Ocelot/Configuration/HttpHandlerOptions.cs:5]:
```suggestion
    /// Describes configuration parameters for http handler, that is created to handle a request to service.
```

File: src/Ocelot/Middleware/DownstreamResponse.cs
Comment (raman-m) [inline src/Ocelot/Middleware/DownstreamResponse.cs:14]:
```suggestion
            Headers = headers ?? new();
```

File: src/Ocelot/Middleware/DownstreamResponse.cs
Comment (raman-m) [inline src/Ocelot/Middleware/DownstreamResponse.cs:45]:
```suggestion
        /// We should make sure we dispose the content and response message to close the connection to the downstream service.
```

File: src/Ocelot/Requester/HttpExceptionToErrorMapper.cs
Comment (raman-m) [inline src/Ocelot/Requester/HttpExceptionToErrorMapper.cs:9]:
```suggestion
        /// <summary>This is a dictionary of custom mappers for exceptions.</summary>
```

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (raman-m) [inline src/Ocelot/Requester/MessageInvokerPool.cs:9]:
```suggestion
    /// <summary>
    /// TODO This should be configurable and available as global config parameter in ocelot.json.
    /// </summary>
```

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (raman-m) [inline src/Ocelot/Requester/MessageInvokerPool.cs:18]: Maybe default factory instance, very simple and lite, is better than exception?

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (raman-m) [inline src/Ocelot/Requester/MessageInvokerPool.cs:94]: Why `private`? Because pooler will work behind the scenes?
Maybe `public` in separate .cs file?!...

File: src/Ocelot/Responder/HttpContextResponder.cs
Comment (raman-m) [inline src/Ocelot/Responder/HttpContextResponder.cs:11]:
```suggestion
/// Cannot unit test things in this class due to methods not being implemented on .NET concretes used for testing.
```

**Review by raman-m — COMMENTED** [review]
If no objections, I will fix styling issue...

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (RaynaldM) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs:83]: this param does not exist in new Ocelot Pipelin

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/MessageInvokerPool.cs:96]: Yes, we could, here it's a struct. I followed an example given by @RaynaldM But, yeah, we could change, easier to unit test.

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs:83]: 😹

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (raman-m) [reply src/Ocelot/Requester/MessageInvokerPool.cs:96]: If it is not hard refactoring, you could move this to separate file while working on tests...

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (raman-m) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: What about this issue?
Does it make sense to have a default factory and removing the exception finally?

Comment (raman-m) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: Excuse me, is `IDelegatingHandlerHandlerFactory` interface registered in DI?

Comment (ggnaegi) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: Yes, as a singleton, the MessageInvokerPool too... By the way, we should have a look at the services lifetime. I can only see singletons, it's a bit odd. But I think we already talked about this.

**Comment by ggnaegi** [comment]
@raman-m so, I have added a memory usage acceptance test in ContentTests. I have tried it with the current develop, and we can see a memory increase of about the payload size. With the current version, it is kept within a range of +/- 10 Mb

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (raman-m) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: Ok, not an issue! Redundant argument null-check is not a problem.
Seems this Pooler service should be shared service aka singleton in DI.
Why do you want review service lifetime?
Does Pooler use shared (static) block of memory? I'm lazy to look into the code...

**Comment by raman-m** [comment]
> @raman-m but we should be careful and foresee **some field tests** when merged to develop, since the refactoring is quite aggressive....

@ggnaegi Yes, it is.
Sorry, I don't understand... What are "some field tests"?

Merging from develop was done by you a couple of hours ago. I'll just press the button, there will be no merge conflicts.

File: src/Ocelot/Requester/MessageInvokerPool.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/MessageInvokerPool.cs:20]: No fine, my bad... But still, I'm a bit surprised, we have Singeltons all over the place.

**Comment by ggnaegi** [comment]
> > @raman-m but we should be careful and foresee **some field tests** when merged to develop, since the refactoring is quite aggressive....
> 
> @ggnaegi Yes, it is. Sorry, I don't understand... What are "some field tests"?
> 
> Merging from develop was done by you a couple of hours ago. I'll just press the button, there will be no merge conflicts.

I mean, wait on feedbacks from @RaynaldM and others before moving to release and create the packages ;-)

**Comment by raman-m** [comment]
Ha, making packages... No! I see you are so hot to release 🤣 
Thwre are the rest of 2 other features in the milestone.
Ok, make sense to wait for Ray's feedback...

**Comment by raman-m** [comment]
Ready to merge! ✔️ 

@ggnaegi @RaynaldM 
Let me know if you agree with merging.

**Comment by IamMartinPeek** [comment]
Hey guys, just asking. When will this be released ? Would need that for my project 👍

**Comment by ggnaegi** [comment]
> Hey guys, just asking. When will this be released ? Would need that for my project 👍

Hello, we have validated the behavior on production environments. It's looking good. we will finalize the release process in the next few days.

**Comment by raman-m** [comment]
@IamMartinPeek Soon... It will be released, I hope, on Monday, Feb 12...
- #1961

**Pull request merged**
